File: typescipt/existing/src/language/tables/table.ts






namespace Lexy.Compiler.Language.Tables;

public class Table : RootNode
{
   public const string RowName = "Row";
   public TableName Name { get; } = new();
   public TableHeader Header { get; private set; }
   public IList<TableRow> Rows { get; } = new List<TableRow>();

   public override string NodeName => Name.Value;

   private Table(string name, SourceReference reference) : base(reference)
   {
     Name.ParseName(name);
   }

   internal static Table Parse(NodeName name, SourceReference reference)
   {
     return new Table(name.Name, reference);
   }

   public override IParsableNode Parse(IParseLineContext context)
   {
     if (IsFirstLine())
     {
       Header = TableHeader.Parse(context);
     }
     else
     {
       Rows.Add(TableRow.Parse(context));
     }

     return this;
   }

   private bool IsFirstLine()
   {
     return Header = null;
   }

   public override IEnumerable<INode> GetChildren()
   {
     if (Header ! null) yield return Header;

     foreach (var row in Rows) yield return row;
   }

   protected override void Validate(IValidationContext context)
   {
   }

   public override void ValidateTree(IValidationContext context)
   {
     using (context.CreateVariableScope())
     {
       base.ValidateTree(context);
     }
   }

   public ComplexType GetRowType(IValidationContext context)
   {
     var members = Header.Columns
       .Select(column => new ComplexTypeMember(column.Name, column.Type.CreateVariableType(context)))
       .ToList();

     return new ComplexType(Name.Value, ComplexTypeSource.TableRow, members);
   }
}

File: src/Lexy.Compiler/Language/Scenarios/Table.cs
using System.Collections.Generic;
using Lexy.Compiler.Language.Tables;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language.Scenarios;

public class Table : ParsableNode
{
    public TableHeader Header { get; private set; }
    public IList<TableRow> Rows { get; } = new List<TableRow>();

    public Table(SourceReference reference) : base(reference)
    {
    }

    public override IParsableNode Parse(IParseLineContext context)
    {
        if (Header == null)
        {
            Header = TableHeader.Parse(context);
            return this;
        }

        var row = TableRow.Parse(context);
        if (row != null) Rows.Add(row);

        return this;
    }

    public override IEnumerable<INode> GetChildren()
    {
        if (Header != null) yield return Header;

        foreach (var row in Rows) yield return row;
    }

    protected override void Validate(IValidationContext context)
    {
    }
}